Date: Apr 17 18:27:42 2017

From: 5bf608725bd2acc2e59dafdf9a797bbc34ff79c6 Jun 06 18:46:17 2017

Subject: [PATCH] DELIA-20408 : Fixing crashes with neigh show

	ip -6 neigh show commands were causing random crashes on xi devices
	Handle use case for IPv6 over IP tunnels.
	This change makes calls to inet_ntop with appropriate flags for v6 mode.

Source: COMCAST


Upstream-Status: Pending

Signed-off-by: skv01 <Shibu_KV@cable.comcast.com>
---


diff --git a/networking/libiproute/libnetlink.c b/networking/libiproute/libnetlink.c
index c7533a4..7aac373 100644
--- a/networking/libiproute/libnetlink.c
+++ b/networking/libiproute/libnetlink.c
@@ -389,6 +389,12 @@ int FAST_FUNC rta_addattr_l(struct rtattr *rta, int maxlen, int type, void *data
 
 void FAST_FUNC parse_rtattr(struct rtattr *tb[], int max, struct rtattr *rta, int len)
 {
+        int index=0;
+        for (index=0;index<max;++index)
+        {
+                tb[index] = NULL;
+        }
+
 	while (RTA_OK(rta, len)) {
 		if (rta->rta_type <= max) {
 			tb[rta->rta_type] = rta;
diff --git a/networking/libiproute/ll_addr.c b/networking/libiproute/ll_addr.c
index 33a54ea..d792f9a 100644
--- a/networking/libiproute/ll_addr.c
+++ b/networking/libiproute/ll_addr.c
@@ -25,7 +25,13 @@ const char* FAST_FUNC ll_addr_n2a(unsigned char *addr, int alen, int type, char
 	) {
 		return inet_ntop(AF_INET, addr, buf, blen);
 	}
+
+        if (alen == 16 && type == ARPHRD_TUNNEL6) {
+                return inet_ntop(AF_INET6, addr, buf, blen);
+        }
+
 	l = 0;
+
 	for (i = 0; i < alen; i++) {
 		if (i == 0) {
 			snprintf(buf + l, blen, ":%02x"+1, addr[i]);
